orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/CompleteExecutionHandler.kt

```diff
@@ -84,11 +84,21 @@ class CompleteExecutionHandler(
           }
         }
       }
-      execution.pipelineConfigId?.let {
-        queue.push(StartWaitingExecutions(it, purgeQueue = !execution.isKeepWaitingPipelines))
+      log.debug("Execution ${execution.id} is with ${execution.status} status and  Disabled concurrent executions is ${execution.isLimitConcurrent}")
+        if (execution.isLimitConcurrent) {
+          if (execution.status != RUNNING) {
+            execution.pipelineConfigId?.let {
+              queue.push(StartWaitingExecutions(it, purgeQueue = !execution.isKeepWaitingPipelines))
+            }
+          } else {
+            log.info("Not starting waiting executions as execution ${execution.id} is currently RUNNING with Disabled concurrent executions.")
+          }
+        } else {
+            log.info("Execution ${execution.id} is not Disabled for concurrent executions, no need to run waiting executions")
```

Comment:
Sorry for the slow updates here - I've been trying to find some time to sit down and read through the context of this change. I'm having a really hard time understanding how this fixes FIFO. 

1. If the pipeline is limiting concurrent executions and this execution isn't RUNNING, push a message to `StartWaitingExecutions` for this pipeline ID.
2. If this execution is RUNNING then don't do anything
3. If this pipeline isn't limiting concurrent executions then don't do anything

I guess my question is why would we have a `CompleteExecution` message at all if the execution is still `RUNNING`?

Reply:
for case 2, even the entire pipeline execution is running and If the pipeline has multiple branches, as per the current behavior  the `CompleteExecution` message is sent on every pipeline stage completion.
Here is the problem once it receives the `CompleteExecution`, it assumes that the execution is complete and start the next waiting execution from the queue.

**Not sure about other impacts If we really stop sending `CompleteExecution` message at all after every stage with branch completion.**

for 3, yes If concurrent execution is enabled, we don't need to run `StartWaitingExecutions`, because it will start immediately and there is no pipelines waiting in queue.

Reply:
Does the first part of the if statement need to push the message too?